### PR TITLE
CI: use `ruff` for linting instead of `flake8`

### PR DESCRIPTION
**Pull request by zacharyburnett**
replace `flake8` with `ruff`

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -85,6 +85,16 @@ namespaces = false
 [tool.pytest.ini_options]
 addopts = "-p no:logging"
 
+[tool.ruff]
+line-length = 99
+extend-ignore = [
+    "E741", # Do not use variables named 'I', 'O', or 'l'
+]
+extend-exclude = [
+    "asv/extern",
```
Comment (HaoZeke) [inline pyproject.toml:94]: This shouldn't be needed, `external` isn't present anymore.

**Review by HaoZeke — COMMENTED** [review]
LGTM, minor nit aside. Does this conflict with `isort` and the other linters? Have you checked that `ruff` can used on all existing files without conflicting with `isort` or the other tools?

**Comment by zacharyburnett** [comment]
> LGTM, minor nit aside. Does this conflict with `isort` and the other linters? Have you checked that `ruff` can used on all existing files without conflicting with `isort` or the other tools?

`ruff` actually has `isort` built into it, though it is disabled by default unless selected. `ruff` doesn't currently implement the `only_sections` parameter however, so kept the `isort` check and it seems to pass the `pre-commit` with no changes.

**Review by HaoZeke — APPROVED** [review]
LGTM, the appveyor seems to be unrelated, could you close/open this so CI runs again?

File: pyproject.toml (hunk 1 above)
Comment (zacharyburnett) [inline pyproject.toml:94]:
```suggestion
```

**Comment by zacharyburnett** [comment]
You could also keep `.flake8` to ensure backwards compatibility with editors that look for it; however, it would require that any changes to the `ruff` config be manually mirrored to `.flake8`

**Comment by mattip** [comment]
I think we should choose one tool as "the way" to format code. The project does not have enough contributors to make installing ruff vs. installing flake8 a big issue. I might have preferred not to touch this at all, but since @zacharyburnett made the effort, let's give it a try.

**Comment by mattip** [comment]
Thanks @zacharyburnett and @HaoZeke

**Pull request merged**